### DISCUSSION TITLE
There is no TimesheetCreateEvent!

### DISCUSSION THREAD

**Discussion by seleondevel** [kimai/kimai#4944] [Plugins]
Kimai has CustomerCreateEvent, ProjectCreateEvent, and ActivityCreateEvent.
But why is there no TimesheetCreateEvent?

For the convenience of writing a timesheet, I would like to fill in the explanation section of the timesheet generation form in advance with the explanation section of Activity, how can I do it?

I can't get timesheet or activity information with TimesheetMetaDefinitionEvent. 

I don't know what to do.

@kevinpapst May I ask for some advice?

**Comment by seleondevel** [comment]
![스크린샷 2024-07-01 182458](https://github.com/kimai/kimai/assets/165639563/4e4e1c31-eeed-4719-9856-529f00bea04d)

**Comment by kevinpapst** [comment]
All available events are here: https://github.com/kimai/kimai/tree/main/src/Event

Read the TimesheetService to find out which events are triggered when.

Creating: https://github.com/kimai/kimai/blob/main/src/Timesheet/TimesheetService.php#L82
Saving: https://github.com/kimai/kimai/blob/main/src/Timesheet/TimesheetService.php#L132

The TimesheetMetaDefinitionEvent is used in many other places and should NOT be used.

You can always register a FormListener to manipulate the TimesheetEditForm.

**Reply by seleondevel** [reply]
@kevinpapst We solved it by putting the following code in the buildForm function of /kimai/src/Form/TimesheetEditForm.php.

$builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event): void {
if(!empty($event->getData()->getActivity())){
$event->getData()->setDescription($event->getData()->getActivity()->getComment());
}
});

However, without modifying the Kimai source code, I would like to implement it in the plug-in I implemented, could you please give me some advice on how to implement the code?

**Reply by seleondevel** [reply]
If I want to use FormEvent for $build in TimesheetEditForm, I have no choice but to add code to /kimai/src/form/TimesheetEditForm.php?

**Reply by seleondevel** [reply]
@kevinpapst When updating Kimai, there are more and more things that need to be modified, so it is burdensome. Do you have any plans to add TimesheetCreateEvent to the next Kimai update version?

**Reply by kevinpapst** [reply]
Then you are doing something wrong. 
You can modify every form from within bundles. 

You can modify every form with a FormExtension and attach a listener to the field:
https://github.com/kimai/kimai/tree/main/src/Form/Extension

This is **NOT** Kimai specific, but plain Symfony knowledge, which I cannot teach.

> Do you have any plans to add TimesheetCreateEvent?

What should that do? Where should it be triggered?

**Reply by seleondevel** [reply]
namespace KimaiPlugin\Bundle\Form\Extension;

use App\Form\TimesheetEditForm;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\FormEvent;
use Symfony\Component\Form\FormEvents;

final class ApiFormExtension extends AbstractTypeExtension
{
    private EntityManagerInterface $entityManager;

    public function __construct(EntityManagerInterface $entityManager)
    {
        $this->entityManager = $entityManager;
    }
    
    public static function getExtendedTypes(): iterable
    {
        return [           
            TimesheetEditForm::class,
        ];
    }    

    public function buildForm(FormBuilderInterface $builder, array $options): void
    { 
        dd('buildForm called');

        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event): void {

            dd('PRE_SET_DATA triggered');
           
        });       

    } 
}


kimai/var/plugins/Bundle/Resources/config/services.yaml

services:
    _defaults:
        autowire: true
        autoconfigure: true
        public: false

    KimaiPlugin\Bundle\:
        resource: '../../*'
        exclude: '../../{Resources}'

    KimaiPlugin\Bundle\Form\Extension\ApiFormExtension:
        arguments:
            $entityManager: '@doctrine.orm.entity_manager'
        tags: ['form.type_extension']

@kevinpapst I implemented the code like this, but the dd function doesn't work. Can you tell me what's wrong?

**Reply by seleondevel** [reply]
@kevinpapst I solved it with TimesheetAdminEditForm, thank you.
